[SDK44] startLocationUpdatesAsync doesn't work when app is killed + Crashes App

### Summary

i'm upgrading from SDK 42 to test the Background location which is already working fine but sometimes crashes with null pointer which is supposed to be fixed here https://github.com/expo/expo/pull/15023 , in changelogs i saw that's been fixed so i tried it on SDK44 . i'm building with the classic build system : expo build


Main Problem Description :  **startLocationUpdatesAsync doesn't work on killed state**

I start the bakground location , it works without a problem on foreground and background .
When i kill the app i'd like to have the location updates for my users who always kill their app .
Anyway : If i kill the App , i wait for while (10 - 20 min ) before the background location updates start , when it does , i see the foreground service but there are no location updates , values are not changing on my server . The same code on SDK42 Works with no problems . I don't know if it's an issue with location module or task-manager.

i attached a mini project to reproduce the Bug .

I hope you can fix this issue ASAP, it was already working fine with SDK42 except for this  issue #15023 
i have a big project based on background locations and i'm having a hard time with this .

Second problem : 

If i kill my app while the background location is on : i get this exception the next time i launch my app , it stucks on splash screen. (to catch this bug the location update must be started , so wait for the updates to start after you kill the app before you launch it again )

`java.lang.RuntimeException: Cannot get ReactInstanceManager from kernel
        at versioned.host.exp.exponent.ExpoTurboPackage.getDefaultUIManager(ExpoTurboPackage.kt:9)
        at versioned.host.exp.exponent.ExpoTurboPackage.getModule(ExpoTurboPackage.kt:3)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:2)
        at com.facebook.react.TurboReactPackage$ModuleHolderProvider.get(TurboReactPackage.java:1)
        at com.facebook.react.bridge.ModuleHolder.create(ModuleHolder.java:8)
        at com.facebook.react.bridge.ModuleHolder.getModule(ModuleHolder.java:7)
        at com.facebook.react.bridge.NativeModuleRegistry.getModule(NativeModuleRegistry.java:11)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:6)
        at com.facebook.react.bridge.CatalystInstanceImpl.getNativeModule(CatalystInstanceImpl.java:1)
        at com.facebook.react.bridge.ReactContext.getNativeModule(ReactContext.java:3)
        at versioned.host.exp.exponent.modules.api.reanimated.NodesManager.<init>(NodesManager.java:13)
        at versioned.host.exp.exponent.modules.api.reanimated.ReanimatedModule.getNodesManager(ReanimatedModule.java:2)
        at versioned.host.exp.exponent.modules.api.reanimated.ReanimatedJSIModulePackage.getJSIModules(ReanimatedJSIModulePackage.java:2)
        at versioned.host.exp.exponent.VersionedUtils$getReactInstanceManagerBuilder$builder$1.getJSIModules(VersionedUtils.kt:6)
        at com.facebook.react.ReactInstanceManager.createReactContext(ReactInstanceManager.java:21)
        at com.facebook.react.ReactInstanceManager.access$1100(ReactInstanceManager.java:1)
        at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:12)
        at java.lang.Thread.run(Thread.java:923)`








[backgroundLocationDemo.zip](https://github.com/expo/expo/files/7697687/backgroundLocationDemo.zip)



### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

44

### Environment

 Expo CLI 4.13.0 environment info:
    System:
      OS: Windows 10 10.0.19042
    Binaries:
      Node: 14.16.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
      npm: 8.1.4 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version     2020.3.0.0 AI-203.7717.56.2031.7583922
    npmPackages:
      expo: ^44.0.0-beta.1 => 44.0.0-beta.1
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    Expo Workflow: managed

### Reproducible demo







[backgroundLocationDemo.zip](https://github.com/expo/expo/files/7697687/backgroundLocationDemo.zip)

@brentvatne  i've created a new project using SDK44 , retested with expo build  just to be sure , i got the same behaviour .

EAS Build :  i gave it a try and i caugh this exception so far : 

`2021-12-13 22:10:49.235 32126-32155/? E/ReactNativeJS: 'TaskManager: Task "locationupdate" failed:', { [ReferenceError: Can't find variable: data] line: 466, column: 2293, sourceURL: 'index.android.bundle' }
2021-12-13 22:10:49.237 1643-2413/? E/JavaBinder: *** Uncaught remote exception!  (Exceptions are not yet supported across processes.)
    java.lang.RuntimeException: android.os.RemoteException: Couldn't get ApplicationInfo for package com.android.accelerometergraph
        at android.os.Parcel.writeException(Parcel.java:2161)
        at android.os.Binder.execTransactInternal(Binder.java:1178)
        at android.os.Binder.execTransact(Binder.java:1123)
     Caused by: android.os.RemoteException: Couldn't get ApplicationInfo for package com.android.accelerometergraph
        at com.android.server.pm.PackageManagerService$PackageManagerNative.getTargetSdkVersionForPackage(PackageManagerService.java:24513)
        at android.content.pm.IPackageManagerNative$Stub.onTransact(IPackageManagerNative.java:255)
        at android.os.Binder.execTransactInternal(Binder.java:1159)
        at android.os.Binder.execTransact(Binder.java:1123) 
2021-12-13 22:10:49.250 32126-32156/? E/unknown:ReactNative: console.error: TaskManager: Task "locationupdate" failed: ReferenceError: Can't find variable: data, stack:
    construct@-1
    p@109:351
    s@107:362
    construct@-1
    <unknown>@106:470
    v@106:658
    h@106:1835
    error@421:631
    <unknown>@477:3077
    y@123:586
    <unknown>@123:1891
    y@123:586
    o@123:1067
    <unknown>@123:1239
    f@119:154
    <unknown>@119:863
    k@125:535
    x@125:933
    callImmediates@125:3022
    value@50:2846
    <unknown>@50:1022
    value@50:2538
    value@50:992
    value@-1
    value@-1`

I'll keep digging to provide max useful data .

--- a/android/expoview/src/main/java/versioned/host/exp/exponent/ExpoTurboPackage.kt
+++ b/android/expoview/src/main/java/versioned/host/exp/exponent/ExpoTurboPackage.kt
@@ -1,7 +1,7 @@
 // Copyright 2020-present 650 Industries. All rights reserved.
 package versioned.host.exp.exponent
 
-import com.facebook.react.CoreModulesPackage
+import com.facebook.react.ReactInstanceManager
 import com.facebook.react.ReactRootView
 import com.facebook.react.TurboReactPackage
 import com.facebook.react.bridge.NativeModule
@@ -21,8 +21,6 @@ import com.facebook.react.uimanager.UIManagerModule
 import com.facebook.react.uimanager.ViewManager
 import com.facebook.systrace.Systrace
 import expo.modules.manifests.core.Manifest
-import host.exp.exponent.di.NativeModuleDepsProvider
-import host.exp.exponent.kernel.Kernel
 import host.exp.exponent.experience.ReactNativeActivity
 import host.exp.exponent.kernel.KernelConstants
 import host.exp.expoview.Exponent
@@ -30,7 +28,6 @@ import versioned.host.exp.exponent.modules.api.reanimated.ReanimatedModule
 import versioned.host.exp.exponent.modules.internal.ExponentAsyncStorageModule
 import versioned.host.exp.exponent.modules.internal.ExponentIntentModule
 import versioned.host.exp.exponent.modules.internal.ExponentUnsignedAsyncStorageModule
-import javax.inject.Inject
 
 /** Package defining basic modules and view managers.  */
 @ReactModuleList(
@@ -46,12 +43,12 @@ class ExpoTurboPackage(
   private val experienceProperties: Map<String, Any?>,
   private val manifest: Manifest
 ) : TurboReactPackage() {
-  @Inject
-  internal lateinit var kernel: Kernel
-
-  init {
-    NativeModuleDepsProvider.instance.inject(ExpoTurboPackage::class.java, this)
-  }
+  // Get the hosted `ReactInstanceManager` by current Activity
+  private val reactInstanceManager: ReactInstanceManager?
+    get() {
+      val currentActivity = Exponent.instance.currentActivity as? ReactNativeActivity ?: return null
+      return (currentActivity.rootView as? ReactRootView)?.reactInstanceManager
+    }
 
   override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
     return listOf()
@@ -70,7 +67,7 @@ class ExpoTurboPackage(
         experienceProperties
       )
       ReanimatedModule.NAME -> ReanimatedModule(context)
-      ReanimatedUIManager.NAME -> createReanimatedUIManager(context) ?: getDefaultUIManager(context)
+      ReanimatedUIManager.NAME -> createReanimatedUIManager(context) ?: throw RuntimeException("Cannot create reanimated uimanager")
       else -> null
     }
   }
@@ -92,11 +89,18 @@ class ExpoTurboPackage(
       val reactModuleInfoMap = mutableMapOf<String, ReactModuleInfo>()
       for (moduleClass in moduleList) {
         val reactModule = moduleClass.getAnnotation(ReactModule::class.java)!!
+        var canOverrideExistingModule = reactModule.canOverrideExistingModule
         val isTurbo = TurboModule::class.java.isAssignableFrom(moduleClass)
+        if (reactModule.name == ReanimatedUIManager.NAME) {
+          if (!shouldOverrideUIManagerForReanimated()) {
+            continue
+          }
+          canOverrideExistingModule = true
+        }
         reactModuleInfoMap[reactModule.name] = ReactModuleInfo(
           reactModule.name,
           moduleClass.name,
-          if (reactModule.name == ReanimatedUIManager.NAME) true else reactModule.canOverrideExistingModule,
+          canOverrideExistingModule,
           reactModule.needsEagerInit,
           reactModule.hasConstants,
           reactModule.isCxxModule,
@@ -116,10 +120,7 @@ class ExpoTurboPackage(
   }
 
   private fun createReanimatedUIManager(reactContext: ReactApplicationContext): UIManagerModule? {
-    val currentActivity = Exponent.instance.currentActivity as? ReactNativeActivity ?: return null
-    val reactInstanceManager = (currentActivity.rootView as? ReactRootView)?.reactInstanceManager
-      ?: return null
-
+    val reactInstanceManager = this.reactInstanceManager ?: return null
     ReactMarker.logMarker(CREATE_UI_MANAGER_MODULE_START)
     Systrace.beginSection(Systrace.TRACE_TAG_REACT_JAVA_BRIDGE, "createUIManagerModule")
     val minTimeLeftInFrameForNonBatchedOperationMs = -1
@@ -135,15 +136,11 @@ class ExpoTurboPackage(
     }
   }
 
-  /**
-   * fallback path to get the react native default UIManagerModule
-   */
-  private fun getDefaultUIManager(context: ReactApplicationContext): UIManagerModule {
-    val reactInstanceManager = kernel.reactInstanceManager
-      ?: throw RuntimeException("Cannot get ReactInstanceManager from kernel")
-    val coreModulesPackage = reactInstanceManager.packages.first { it is CoreModulesPackage } as? CoreModulesPackage
-      ?: throw RuntimeException("Cannot get CoreModulesPackage")
-    return coreModulesPackage.getModule(UIManagerModule.NAME, context) as UIManagerModule
+  // Reanimated UIManager requires `ReactInstanceManager`,
+  // for headless mode we cannot get the hosted `ReactInstanceManager` from current Activity,
+  // in this case, we don't override UIManager for reanimated.
+  private fun shouldOverrideUIManagerForReanimated(): Boolean {
+    return this.reactInstanceManager != null
   }
 
   companion object {

--- a/android/versioned-abis/expoview-abi44_0_0/src/main/java/abi44_0_0/host/exp/exponent/ExpoTurboPackage.kt
+++ b/android/versioned-abis/expoview-abi44_0_0/src/main/java/abi44_0_0/host/exp/exponent/ExpoTurboPackage.kt
@@ -1,7 +1,7 @@
 // Copyright 2020-present 650 Industries. All rights reserved.
 package abi44_0_0.host.exp.exponent
 
-import abi44_0_0.com.facebook.react.CoreModulesPackage
+import abi44_0_0.com.facebook.react.ReactInstanceManager
 import abi44_0_0.com.facebook.react.ReactRootView
 import abi44_0_0.com.facebook.react.TurboReactPackage
 import abi44_0_0.com.facebook.react.bridge.NativeModule
@@ -21,8 +21,6 @@ import abi44_0_0.com.facebook.react.uimanager.UIManagerModule
 import abi44_0_0.com.facebook.react.uimanager.ViewManager
 import abi44_0_0.com.facebook.systrace.Systrace
 import expo.modules.manifests.core.Manifest
-import host.exp.exponent.di.NativeModuleDepsProvider
-import host.exp.exponent.kernel.Kernel
 import host.exp.exponent.experience.ReactNativeActivity
 import host.exp.exponent.kernel.KernelConstants
 import host.exp.expoview.Exponent
@@ -30,7 +28,6 @@ import abi44_0_0.host.exp.exponent.modules.api.reanimated.ReanimatedModule
 import abi44_0_0.host.exp.exponent.modules.internal.ExponentAsyncStorageModule
 import abi44_0_0.host.exp.exponent.modules.internal.ExponentIntentModule
 import abi44_0_0.host.exp.exponent.modules.internal.ExponentUnsignedAsyncStorageModule
-import javax.inject.Inject
 
 /** Package defining basic modules and view managers.  */
 @ReactModuleList(
@@ -46,12 +43,12 @@ class ExpoTurboPackage(
   private val experienceProperties: Map<String, Any?>,
   private val manifest: Manifest
 ) : TurboReactPackage() {
-  @Inject
-  internal lateinit var kernel: Kernel
-
-  init {
-    NativeModuleDepsProvider.instance.inject(ExpoTurboPackage::class.java, this)
-  }
+  // Get the hosted `ReactInstanceManager` by current Activity
+  private val reactInstanceManager: ReactInstanceManager?
+    get() {
+      val currentActivity = Exponent.instance.currentActivity as? ReactNativeActivity ?: return null
+      return (currentActivity.rootView as? ReactRootView)?.reactInstanceManager
+    }
 
   override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
     return listOf()
@@ -70,7 +67,7 @@ class ExpoTurboPackage(
         experienceProperties
       )
       ReanimatedModule.NAME -> ReanimatedModule(context)
-      ReanimatedUIManager.NAME -> createReanimatedUIManager(context) ?: getDefaultUIManager(context)
+      ReanimatedUIManager.NAME -> createReanimatedUIManager(context) ?: throw RuntimeException("Cannot create reanimated uimanager")
       else -> null
     }
   }
@@ -92,11 +89,18 @@ class ExpoTurboPackage(
       val reactModuleInfoMap = mutableMapOf<String, ReactModuleInfo>()
       for (moduleClass in moduleList) {
         val reactModule = moduleClass.getAnnotation(ReactModule::class.java)!!
+        var canOverrideExistingModule = reactModule.canOverrideExistingModule
         val isTurbo = TurboModule::class.java.isAssignableFrom(moduleClass)
+        if (reactModule.name == ReanimatedUIManager.NAME) {
+          if (!shouldOverrideUIManagerForReanimated()) {
+            continue
+          }
+          canOverrideExistingModule = true
+        }
         reactModuleInfoMap[reactModule.name] = ReactModuleInfo(
           reactModule.name,
           moduleClass.name,
-          if (reactModule.name == ReanimatedUIManager.NAME) true else reactModule.canOverrideExistingModule,
+          canOverrideExistingModule,
           reactModule.needsEagerInit,
           reactModule.hasConstants,
           reactModule.isCxxModule,
@@ -116,10 +120,7 @@ class ExpoTurboPackage(
   }
 
   private fun createReanimatedUIManager(reactContext: ReactApplicationContext): UIManagerModule? {
-    val currentActivity = Exponent.instance.currentActivity as? ReactNativeActivity ?: return null
-    val reactInstanceManager = (currentActivity.rootView as? ReactRootView)?.reactInstanceManager
-      ?: return null
-
+    val reactInstanceManager = this.reactInstanceManager ?: return null
     ReactMarker.logMarker(CREATE_UI_MANAGER_MODULE_START)
     Systrace.beginSection(Systrace.TRACE_TAG_REACT_JAVA_BRIDGE, "createUIManagerModule")
     val minTimeLeftInFrameForNonBatchedOperationMs = -1
@@ -135,15 +136,11 @@ class ExpoTurboPackage(
     }
   }
 
-  /**
-   * fallback path to get the react native default UIManagerModule
-   */
-  private fun getDefaultUIManager(context: ReactApplicationContext): UIManagerModule {
-    val reactInstanceManager = kernel.reactInstanceManager
-      ?: throw RuntimeException("Cannot get ReactInstanceManager from kernel")
-    val coreModulesPackage = reactInstanceManager.packages.first { it is CoreModulesPackage } as? CoreModulesPackage
-      ?: throw RuntimeException("Cannot get CoreModulesPackage")
-    return coreModulesPackage.getModule(UIManagerModule.NAME, context) as UIManagerModule
+  // Reanimated UIManager requires `ReactInstanceManager`,
+  // for headless mode we cannot get the hosted `ReactInstanceManager` from current Activity,
+  // in this case, we don't override UIManager for reanimated.
+  private fun shouldOverrideUIManagerForReanimated(): Boolean {
+    return this.reactInstanceManager != null
   }
 
   companion object {

